zebra-chain/src/subtree.rs

```diff
@@ -0,0 +1,64 @@
+//! Struct representing Sapling/Orchard note commitment subtrees
+
+use std::sync::Arc;
+
+use crate::block::Height;
+
+/// Height at which Zebra tracks subtree roots
+pub const TRACKED_SUBTREE_HEIGHT: u8 = 16;
+
+/// Size of tracked subtrees
+pub const TRACKED_SUBTREE_SIZE: u64 = 1 << TRACKED_SUBTREE_HEIGHT;
+
+/// A subtree index
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct NoteCommitmentSubtreeIndex(pub u16);
+
+impl From<u16> for NoteCommitmentSubtreeIndex {
+    fn from(value: u16) -> Self {
+        Self(value)
+    }
+}
+
+/// Subtree of Sapling or Orchard note commitment tree
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct NoteCommitmentSubtree<Node> {
+    /// Index of this subtree
+    pub index: NoteCommitmentSubtreeIndex,
+    /// End boundary of this subtree, the block height of its last leaf.
+    pub end: Height,
+    /// Root of this subtree.
+    pub node: Node,
+}
+
+impl<Node> NoteCommitmentSubtree<Node> {
+    /// Creates new [`NoteCommitmentSubtree`]
+    pub fn new(index: impl Into<NoteCommitmentSubtreeIndex>, end: Height, node: Node) -> Arc<Self> {
+        let index = index.into();
+        Arc::new(Self { index, end, node })
+    }
+}
+
+/// Partial subtree of Sapling note commitment tree
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct PartialNoteCommitmentSubtree<Node> {
```

Comment:
```suggestion
/// Subtree root of Sapling or Orchard note commitment tree, with block height, but without the subtree index.
/// Used for database key-value serialization, where the subtree index is the key, and this struct is the value.
#[derive(Copy, Clone, Debug, Eq, PartialEq)]
pub struct NoteCommitmentSubtreeData<Node> {
```

Reply:
This rename should be scheduled after the existing code has merged. I'll open a ticket for cleanups if it turns out to be a priority.

The doc change can be applied in this PR though.